Technical documentation about the namespace replication mechanism (#831)
liqotech/liqo: docs/pages/concepts/offloading/namespace-replication.md

---
title: Namespace replication
weight: 2
---

### The Liqo Namespace Model

> The Liqo resource replication model is composed of several concurring logics responsible for different kind of objects to replicate.

Reviewer: ```suggestion
The Liqo resource replication model is made by a logic that replicate objects with a behavior that is object-specific.
```

One important example is represented by namespaces. 
In particular, namespace replication across multiple clusters implements a key mechanism to extend seamleassly a cluster to some others.

> In the Liqo model, the namespace replication corresponds to a standard v1.Namespace and an associated [NamespaceOffloading](#) CR.

Reviewer: ```suggestion
In the Liqo model, the namespace replication corresponds to a standard `v1.Namespace` and an associated [NamespaceOffloading](#) CR.
```

The NamespaceOffloading resource allows to select the clusters which are the clusters where the application contained inside the namespace can be scheduled.

> In [a dedicated usage section](/usage/namespace_offloading), you can learn how to define and use a NamespaceOffloading resource activating the replication of the local namespace on peered clusters.

Reviewer: ```suggestion
Section [Namespace Extensions](/usage/namespace_offloading) shows how to define and use a `NamespaceOffloading` resource that enables the replication of the local namespace on peered (remote) clusters.
```

> In this section, you may discover how Liqo namespace model works under the hood, undestanding which resources and controllers are activated when a new namespaceOffloading is manipulated.

Reviewer: ```suggestion
This section shows how the Liqo namespace model works under the hood, understanding which resources and controllers are activated when a new `namespaceOffloading` is manipulated.
```

#### Resources

Liqo namespace replication involves instances of two kinds of CRDs:

* First, the initial trigger is represented by the manipulation of a **NamespaceOffloading** (or the labelling of a namespace as explained in [the usage section](/usage/namespace_offloading)).
  On the one hand, the *NamespaceOffloading* spec describes the properties of the replication, such as the name of the replicated namespaces.

>   On the other hand, the status collects the information about the actual status of remote namespaces (i.e. if the replication has succeeded).

Reviewer: ```suggestion
  On the other hand, the `status` collects the information about the actual status of remote namespaces, e.g., if the replication succeeded or not.
```

> * Second, **NamespaceMap** resources contain the list of namespaces associated to a specific node.

Reviewer: ```suggestion
* **NamespaceMap**: this CR contains the list of namespaces associated to a specific node.
```

Reviewer: I don't know what "list of namespaces associated to a specific node" means.

Author: I have tried to clarify the concept, thank you.

>   The spec collects the list of desired namespaces for a specific remote cluster while the status updated information about their effective creation.

Reviewer: ```suggestion
  The `spec` collects the list of desired namespaces for a specific remote cluster while the `status` keeps the updated information about their actual creation.
```

>   Each *NamespaceMap* can be alimented by several *NamespaceOffloading* instances that are targeting the same cluster for a remote namespace.

Reviewer: Not clear this sentence. It could be useful to continue with an example.
E.g., "For example, ..."

Author: Sure, I added it, thank you.

>   It is worth noting that the NamespaceMap status represent the source of truth to know the status of replicated namesapces on foreign clusters.

Author: Custom resources and resource fields must be both written in Pascal code: NamespaceMap, CurrentMapping, etc...
We do not use the backticks `NamespaceMap` because the actual Hugo theme, inserts a light yellow box with the copy and paste functionality, it is not like there in GitHub. 
For standard k8s words like status, spec, etc... we usually use lowercase letters as usual.

#### Controllers

> The namespace replication logic is composed of several distinct controllers, responsible for different aspects of namespace controllers:

Reviewer: ```suggestion
The namespace replication logic is made by several controllers, responsible for different aspects of namespace controllers:
```

> * **NamespaceOffloading Controller**: processes the NamespaceOffloading spec and outputs namespace entries to NamespaceMap specs.

Reviewer: ```suggestion
* **NamespaceOffloading Controller**: it processes the NamespaceOffloading spec and creates the proper namespace entries into NamespaceMap specs.
```

> * **NamespaceMap Controller**: enforces the namespace creation of the remote namespaces on remote clusters and updates the status of the NamespaceMaps.

Reviewer: ```suggestion
* **NamespaceMap Controller**: it takes care of creating the required namespaces on remote clusters and updates their corresponding status in the NamespaceMaps.
```

> * **OffloadingStatus Controller**: updates the status of the NamespaceOffloading resources by reading the NamespaceMaps statuses.

Reviewer: ```suggestion
* **OffloadingStatus Controller**: it updates the status of the NamespaceOffloading resources by reading the NamespaceMaps statuses.
```

Reviewer: I did not understand exactly what this mean.

Author: We have fixed that, thank you

#### Workflow

> In the next figure, you can observe a representation of the overall workflow steps of the replication process:

Reviewer: ```suggestion
The figure below presents a representation of the overall workflow steps of the replication process:
```

> ![](/images/namespace-replication/replication.png)

Reviewer: This picture is very good, but is very difficult to understand without an example.
I would suggest to duplicate this picture: a first part contains the "theory", such as now;a second part contains a simple, very simple example, which could be referenced in the explanation steps below.

Author: I added an example figure, let me know if it looks good to you.

We can resume the steps in:

1. When the user creates/updates a **NamespaceOffloading** object in a Liqo-enabled namespace **(Step 1 in the figure)**, the Liqo logic processes the resource spec.

> 2. After having detected the virtual-nodes compliant with selector, a namespace creation request is entered in every NamespaceMap associated with a selected cluster **(Step 2 in the figure)**.

Reviewer: Please start this sentence with a subject. E.g.:
Component XX, after having bla bla, will do ble ble

In particular, it considers the **ClusterSelector** field to select on which cluster the current namespace should be replicated.

> 

Reviewer: ```suggestion
```

More precisely, the controller responsible for this reconciliation is the **NamespaceOffloading Controller**. 
It processes the *NamespaceOffloading* spec fields, by inserting the namespace creation requests in the *spec.DesiredMapping* field of **NamespaceMap** instances of selected clusters.
The request format is an entry consisting of the local namespace name as a key and the name of the remote namespace as a value.
In addition, this operation has to be performed every time a new virtual-node joins the cluster after a new peering is established.

> 3. Once a *NamespaceMap* is updated, it is time to create the corresponding remote namespace. 

Reviewer: Same comment as item (2).

The *NamespaceMap Controller* reconciles the NamespaceMap resources, creating the remote namespaces and storing the operation results in the corresponding NamespaceMap.
More precisely, the occurred creation is saved in the *status.CurrentMapping* field, as you can see in **Step 3 of the figure**.
The result format is similar to request one: the key is the name of the local namespace, while the value is composed of the remote name and the actual remote namespace phase.

> In addition to the actual creation of namespaces, the *NamespaceMap Controller* (1) periodically checks that each entry in the *spec.desiredMapping* field has an associated remote namespace and (2) performs health checks on the remote namespaces. The result of all its operations are stored in the *status.CurrentMapping* field in the NamespaceMap resource.

Reviewer: Is this step (4)?
Otherwise, step (4) is missing.

Author: We added it, thank you.

5. Liqo periodically checks that the requested remote namespaces are present.
Whenever it detects a change in the namespaces state, it immediately updates the NamespaceMap resources.
The NamespaceOffloading status is updated thanks to NamespaceMap status changes **(step 4 in the figure)**.
The **OffloadingStatus Controller** is responsible for this NamespaceOffloading status reconciliation. It periodically checks the status of all *NamespaceMaps* in the clusters, and for each NamespaceOffloading object, (1) it updates the RemoteNamespaceConditions with the actual remote namespaces status and (2)  it changes the global OffloadingPhase according to the previously set remote conditions. 

As already detailed in the [NamespaceOffloading status description](/usage/namespace_offloading/#check-the-namespaceoffloading-resource-status), the fields that provide the user with all the information about the replication phase are the RemoteNamespaceConditions and the OffloadingPhase **(step 5 in the figure)**. 

> ### Deletion workflow

Reviewer: Transform this text into a set of steps, as before.
